core/trino-main/src/main/java/io/trino/dispatcher/QueuedStatementResource.java

```diff
@@ -400,15 +440,86 @@ public QueryResults getQueryResults(long token, ExternalUriInfo externalUriInfo)
 
         public void cancel()
         {
-            creationFuture.addListener(() -> dispatchManager.cancelQuery(queryId), directExecutor());
+            cancelled = true;
+            creationFuture.addListener(() -> cancelInternal(), directExecutor());
+        }
+
+        private void cancelInternal()
+        {
+            if (!dispatchManager.cancelQuery(queryId)) {
```

Comment:
why can't we have event logic in `dispatchManager.cancelQuery`? 
Is that lack of session or the fact that is used from QueryResource too? Or ... ?